rmw_cyclonedds_cpp/src/rmw_node.cpp

```diff
@@ -2812,41 +2826,268 @@ extern "C" rmw_ret_t rmw_get_node_names(
+    node_ns = "_NODE_NAMESPACE_UNKNOWN_";
+  }
+  return RMW_RET_OK;
+}
+
+static rmw_ret_t
+set_rmw_topic_endpoint_info(
+  node_cache_t & node_cache,
+  dds_entity_t rd,
+  rcutils_allocator_t * allocator,
+  const endpoint_info_t & epinfo,
+  bool is_publisher,
+  rmw_topic_endpoint_info_t & topic_endpoint_info)
+{
+  static_assert(
+    sizeof(dds_builtintopic_guid_t) <= RMW_GID_STORAGE_SIZE,
+    "RMW_GID_STORAGE_SIZE insufficient to store the rmw_cyclonedds_cpp GID implementation."
+  );
+  rmw_ret_t ret;
+  std::string node_name;
+  std::string node_ns;
+  const auto it = node_cache.find(epinfo.node_key);
+  if (it != node_cache.end()) {
+    node_name = it->second.first;
+    node_ns = it->second.second;
+  } else {
+    if ((ret =
+      get_node_name(rd, epinfo.node_key, epinfo.node_insth, node_name, node_ns)) != RMW_RET_OK)
+    {
+      return ret;
+    }
+    node_cache[epinfo.node_key] = std::make_pair(node_name, node_ns);
+  }
+  if ((ret = rmw_topic_endpoint_info_set_node_name(
+      &topic_endpoint_info,
+      node_name.c_str(),
+      allocator)) != RMW_RET_OK)
+  {
+    return ret;
+  }
+  if ((ret = rmw_topic_endpoint_info_set_node_namespace(
+      &topic_endpoint_info,
+      node_ns.c_str(),
+      allocator)) != RMW_RET_OK)
+  {
+    return ret;
+  }
+  if ((ret = rmw_topic_endpoint_info_set_topic_type(
+      &topic_endpoint_info,
+      epinfo.topic_type.c_str(),
+      allocator)) != RMW_RET_OK)
+  {
+    return ret;
+  }
+  if ((ret = rmw_topic_endpoint_info_set_endpoint_type(
+      &topic_endpoint_info,
+      is_publisher ? RMW_ENDPOINT_PUBLISHER : RMW_ENDPOINT_SUBSCRIPTION)) != RMW_RET_OK)
+  {
+    return ret;
+  }
+  if ((ret = rmw_topic_endpoint_info_set_gid(
+      &topic_endpoint_info,
+      epinfo.endpoint_key.v,
+      sizeof(dds_builtintopic_guid_t))) != RMW_RET_OK)
+  {
+    return ret;
+  }
+  topic_endpoint_info.qos_profile = epinfo.qos_profile;
+
+  return RMW_RET_OK;
+}
+
+static void handle_topic_endpoint_info_fini(
+  rmw_topic_endpoint_info_t * topic_endpoint_info,
+  rcutils_allocator_t * allocator)
+{
+  std::string err_msg;
+  rmw_ret_t ret;
+  bool err_set = rmw_error_is_set();
+  if (err_set) {
+    err_msg = rmw_get_error_string().str;
+  }
+  rmw_reset_error();
+  if ((ret = rmw_topic_endpoint_info_fini(topic_endpoint_info, allocator)) != RMW_RET_OK) {
+    RCUTILS_LOG_ERROR_NAMED(
+      "rmw_cyclonedds_cpp",
+      "rmw_topic_endpoint_info_fini failed: %s",
+      rmw_get_error_string().str);
+    rmw_reset_error();
+  }
+  if (err_set) {
+    RMW_SET_ERROR_MSG(err_msg.c_str());
+  }
+}
+
+static rmw_ret_t
+get_endpoint_info_by_topic(
+  const rmw_node_t * node,
+  rcutils_allocator_t * allocator,
+  const char * topic_name,
+  bool no_mangle,
+  bool is_publisher,
+  rmw_topic_endpoint_info_array_t * endpoint_info)
+{
+  auto node_impl = static_cast<CddsNode *>(node->data);
+  const auto re_typ = std::regex("^(.*::)dds_::(.*)_$", std::regex::extended);
```

Comment:
I think it would make sense to create a function to demangle topic names/types, with a parameter specifying whether it concerns data, requests or responses. There was some duplication already, but this adds yet another one and having regexes all over the place is a bit unpleasant.